### PR TITLE
Implement token exchange

### PR DESCRIPTION
**Pull request by angelakis**
**Update**: 
We decided to refactor token coop based on the by_grant_type branch. Check https://github.com/IdentityPython/oidcendpoint/pull/59#issuecomment-689652183 for more info.

**Original Message:**
This is a pull request to implement token exchange a.k.a. [RFC 8693](https://tools.ietf.org/html/rfc8693).

The implementation will not be as straightforward as I first hoped it would be, as token exchange is an extension grant type used in the token endpoint.

Currently implemented token and tokencoop endpoints, do not have an easy way to add/configure new grant_types so I think we should refactor one of those two.

TokenCoop was selected to be refactored but we could in the future
just keep the new version instead of having both Token and the new
configurable/extendable TokenCoop.

I think it would be a good idea to use grant_types_supported as a configuration option to figure out what grant types the endpoint should support.

As I previously mentioned, this is still a work in progress and the actual token exchange code has not yet been implemented. RFC 8693 is broad so its implementation will need a discussion too.

Any feedback welcome.

### CODE REVIEW

File: src/oidcendpoint/oidc/token_coop.py [hunk 1]
```diff
@@ -165,6 +172,14 @@ def _access_token_post_parse_request(self, request, client_id="", **kwargs):
 
         return request
 
+    def check_grant_type_allowed(self, grant_type):
+        grant_types_supported = self.endpoint_context.provider_info['grant_types_supported']
+        if (
+            grant_type not in self.token_grant_types
+            or grant_type not in grant_types_supported
+        ):
+            raise ProcessError("Invalid grant type")
```
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: Add the `grant_type` value to the exception?

File: src/oidcendpoint/oidc/token_coop.py [hunk 2]
```diff
@@ -196,14 +211,48 @@ def _refresh_token_post_parse_request(self, request, client_id="", **kwargs):
 
         return request
 
+    def _token_exchange_post_parse_request(self, request, client_id="", **kwargs):
```
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: We could override the `parse_request` method, add a `request_cls` param to override `self.request_cls` and provide the proper one depending on the grant_type. This way we can get rid of `*_post_parse_request`

File: src/oidcendpoint/oidc/token_coop.py [hunk 3]
```diff
@@ -196,14 +211,48 @@ def _refresh_token_post_parse_request(self, request, client_id="", **kwargs):
...
-            if self.allow_refresh:
-                return self._refresh_token_post_parse_request(request, client_id, **kwargs)
-            else:
-                raise ProcessError("Refresh Token not allowed")
+        elif grant_type == "refresh_token":
+            return self._refresh_token_post_parse_request(request, client_id, **kwargs)
+        elif grant_type == "urn:ietf:params:oauth:grant-type:token-exchange":
+            return self._token_exchange_post_parse_request(request, client_id, **kwargs)
+        else:
+            raise ProcessError("Invalid grant type")
```
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: Add the `grant_type` value to the exception?

File: src/oidcendpoint/oidc/token_coop.py [hunk 4]
```diff
@@ -214,13 +263,24 @@ def process_request(self, request=None, **kwargs):
...
+            if grant_type == "authorization_code":
                 logger.debug("Access Token Request")
                 response_args = self._access_token(request, **kwargs)
-            elif request["grant_type"] == "refresh_token":
+            elif grant_type == "refresh_token":
                 logger.debug("Refresh Access Token Request")
                 response_args = self._refresh_access_token(request, **kwargs)
+            elif grant_type == "urn:ietf:params:oauth:grant-type:token-exchange":
+                logger.debug("Token Exchange Request")
+                response_args = self._token_exchange(request, **kwargs)
```
Comment (nsklikas) [inline src/oidcendpoint/oidc/token_coop.py]: Move this `if/elif/elif` to separate function?

File: src/oidcendpoint/oidc/token_coop.py (hunk 1 above)
Comment (rohe) [reply src/oidcendpoint/oidc/token_coop.py]: yes

File: src/oidcendpoint/oidc/token_coop.py (hunk 2 above)
Comment (rohe) [reply src/oidcendpoint/oidc/token_coop.py]: I'd rather add a get_request_class() method.

File: src/oidcendpoint/oidc/token_coop.py (hunk 3 above)
Comment (rohe) [reply src/oidcendpoint/oidc/token_coop.py]: yes

**Comment by rohe** [comment]
Have a look at https://github.com/IdentityPython/oidcendpoint/blob/by_grant_type/src/oidcendpoint/oidc/token_coop.py
I'd rather do it that way.
Now, what you see there is just a first iteration. It should of course be configurable.
But you can see the layout.

**Comment by rohe** [comment]
I've even added a test based on an example in the RFC.
It's extremely crude but it works :-)

**Comment by rohe** [comment]
And yes, I think we should get rid of Token and only keep TokenCoop. 
Once we've done that we should probably rename TokenCoop to Token.

**Comment by angelakis** [comment]
We decided to use @rohe's  by_grant_type branch and work on that to implement a simple Token Exchange functionality.

I believe that our first goal should be to support exchanging access tokens for new access tokens with (the same or) altered scopes. 

In that regard I have just pushed the current implementation which seems to work.

There's one simple change concerning by_grant_type and its configuration:
                                                  
    The user can now set the usual grant types to have a default                  
    configuration and class, without the need to set the oidcendpoint             
    class path. This can be done by setting a value of "default" or               
    None (no value in yaml dict).

There are multiple issues right now and the main one is probably how session works, and its dependency on state. We should discuss a way forward concerning that.

Tests are also missing currently.

Any other feedback welcome.

**Comment by nsklikas** [comment]
Can you clean up the commits a little bit? Also it seems that there are extra changes in this PR that are not related to token exchange.

File: src/oidcendpoint/client_authn.py [hunk 5]
```diff
@@ -386,12 +390,13 @@ def verify_client(
...
     if client_id:
-        if not client_id in endpoint_context.cdb:
+        _cinfo = endpoint_context.cdb.get(client_id)
+        if _cinfo is None:
             raise UnknownClient("Unknown Client ID")
 
-        _cinfo = endpoint_context.cdb[client_id]
-        if isinstance(_cinfo, str):
-            if not _cinfo in endpoint_context.cdb:
+        if isinstance(_cinfo, str): # This is if _cinfo is a reference to another client_id
```
Comment (nsklikas) [inline src/oidcendpoint/client_authn.py]: Why is this needed?

**Comment by angelakis** [comment]
Pushed multiple tests both for the by-grant-type refactor and token exchange. They try to highlight the functionality we introduce so they will need changes once we decide on the token exchange features (or whether we allow specifying "default" on a grant_type in grant_types_supported).

**Comment by angelakis** [comment]
> Can you clean up the commits a little bit? Also it seems that there are extra changes in this PR that are not related to token exchange.

I could remove the isort commits and keep only the parts containing the actual changes in TokenCoop, if that's ok with @rohe. And if we want to do this even better I could split up the grant_types_supported changes and the token exchange ones.

**Comment by rohe** [comment]
That would be OK with me.

**Comment by angelakis** [comment]
I cleaned up unrelated changes and then split up the changes in two different branches, creating #88 with the configurable-grant-types parts. So only token exchange related changes should remain here.

This branch is now based on #88 and should be merged after that.

**Comment by angelakis** [comment]
Accidentally closed this while trying to push the rebased version to the new session handling. Seeing that this PR has changed a lot since its creation, it may be better to create a new one.
